src/security/KeyData.cc

```diff
@@ -93,45 +95,48 @@ Security::KeyData::loadX509ChainFromFile()
         return;
     }
 
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-    if (X509_check_issued(cert.get(), cert.get()) == X509_V_OK) {
-        char *nameStr = X509_NAME_oneline(X509_get_subject_name(cert.get()), nullptr, 0);
-        debugs(83, DBG_PARSE_NOTE(2), "Certificate is self-signed, will not be chained: " << nameStr);
-        OPENSSL_free(nameStr);
-    } else
-#endif
-    {
-        debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
-        // and add to the chain any other certificate exist in the file
-        CertPointer latestCert = cert;
-
-        while (const auto ca = Ssl::ReadX509Certificate(bio)) {
-            // get Issuer name of the cert for debug display
-            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca.get()), nullptr, 0);
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (X509_check_issued(ca.get(), ca.get()) == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
-                OPENSSL_free(nameStr);
-                continue;
-            }
-#endif
-            // checks that the chained certs are actually part of a chain for validating cert
-            const auto checkCode = X509_check_issued(ca.get(), latestCert.get());
-            if (checkCode == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
-                // OpenSSL API requires that we order certificates such that the
-                // chain can be appended directly into the on-wire traffic.
-                latestCert = CertPointer(ca);
-                chain.emplace_back(latestCert);
-            } else {
-                debugs(83, DBG_PARSE_NOTE(2), certFile << ": Ignoring non-issuer CA " << nameStr << ": " << X509_verify_cert_error_string(checkCode) << " (" << checkCode << ")");
+    CertList intermediates;
+    debugs(83, DBG_PARSE_NOTE(3), "Building certificate chain from " << certFile);
+    while (const auto ca = Ssl::ReadX509Certificate(bio)) {
```

Comment:
> We call this list `intermediates` but it actually contains the signing/leaf certificate as well because `certFile` is where the signing certificate is stored, right?  If so, then I think we should not add the signing certificate to `intermediates`.

Yes but the signing certificate, if it is not a root CA (self signed) it is an intermediate certificate.

>  I suspect the signing certificate will not be added to the `chain` in the nested loop further below anyway, but I think that adding a non-intermediate to `intermediates` is dangerously misleading.

It will not be added to the chain. But the signing certificate if it is not a root certificate is an intermediate.

The problem with KeyData is that it is used:
* For standard HTTPS proxying where we must not chain the loaded certificate (the 1st certificate in file). This is the server certificate
* For bumping http ports. In this case the loaded certificate must be chained if it is intermediate certificate, because it is used to sign the generated certificate.
* The ServerOptions::SingingCa is not well named because in some cases is a singing CA but in other cases (HTTPS proxying) is just a server certificate.
* To make thinks worst for bumping case, always created a staticContext which uses the singingCa as primary cert. This is used when the ServerOptions::generateHostCertificates is false.

In practice the signingCA should be added in `chain` list in the cases it should be chained with other certs. But as is the squid code now it is complex to do it correctly. So, in the current squid code, the decision to chain or not the SingingCa is done after we are generating a certificate. 

I tried some changes in order to add to `chain` list exactly the certificates must chained but I scared a little.
I left as is. But I can retry if you think this is required.

> 
> BTW, why adding cert to `intermediates` does not trigger "Unused intermediate certificate" warnings?

I am not understanding the question. Why it should trigger this warning in this case?

But I found a bug here, always this error is appeared for signing certificate. 
This bug should fixed with commit 03adb8151

> 
> Moreover, it looks like we are opening and loading certFile twice, once in loadX509CertFromFile() and once here, in loadX509ChainFromFile(). I am pretty sure we can simplify/clarify our code if we merge those two methods together, opening the file once, and loading each certificate there once.

I agree. This is fixed with commit 03adb8151

Reply:
> > BTW, why adding cert to `intermediates` does not trigger "Unused intermediate certificate" warnings?

> I am not understanding the question. Why it should trigger this warning in this case?

... because "cert" (i.e. the signing certificate in this context; the same certificate as the one pointed to by the `cert` field) is going to be considered "unused" in this scope (i.e. it is not used to sign any other certificate; it is a leaf certificate used for signing traffic, not signing certificates) and, hence, `cert` will remain in `intermediates` after the loop, generating a warning. I think you have actually confirmed that yourself:

> But I found a bug here, always this error is appeared for signing certificate.

Since "cert" in my question _is_ the (copy of the signing certificate `cert`), you have confirmed my fears about the warnings and effectively answered the question -- yes, the warning is going to be triggered.

> This bug should fixed with commit https://github.com/measurement-factory/squid/commit/03adb815142419a62cf9c7c405ba2f5ab3f2d2a7

That commit went into a different branch (SQUID-642-Sending-cientca-certificates -- an old deprecated branch). Was it meant to be on this PR branch instead (SQUID-642-do-not-chain-self-signed-certs)? Please do not copy/apply that commit to this or any other branch yet. I am just trying to understand what happened _before_ we modify something and increase my confusion.

I agree that 03adb815142419a62cf9c7c405ba2f5ab3f2d2a7 fixes the bug I was worried about here because that commit stops adding the (copy of the) signing certificate to the (now renamed) `intermediates`.